examples/other/image_histogram.py

```diff
@@ -0,0 +1,582 @@
+    let channel = global_invocation_id.y;  // 0..4
+
+    let binValue = s_bins[i* 4 + channel];
+
+    // The line consists of 4 pieces, one for each channel.
+    // The '+ channel' is there because we need to put a nan-vertex in between
+    // the line pieces.
+    let vertex_index = ((channel * binCount) + channel + i) * 3u;
+
+    // Write a nan value between line pieces.
+    if (i == 0u && channel > 0u) {
+        let nan = bitcast<f32>(0x7fc00000u);  // nan;
+        s_positions[vertex_index-3] = nan;
+        s_positions[vertex_index-2] = nan;
+        s_positions[vertex_index-1] = nan;
+    }
+
+    // Write the value
+    let ix = vertex_index;
+    let iy = vertex_index + 1;
+    let iz = vertex_index + 2;
+    s_positions[ix] = f32(i);
+    s_positions[iy] = f32(binValue) * scale;
+    s_positions[iz] = 0.0;
+}
+
+
+"""
+
+
+hist_calc_shader = ComputeShader(
+    histogram_wgsl,
+    entry_point="calc_histogram",
+    report_time=True,
+)
+hist_calc_shader.set_resource(0, image_object.geometry.grid)
+hist_calc_shader.set_resource(1, histogram_bins_buffer, clear=True)
+
+hist_write_shader = ComputeShader(
+    histogram_wgsl,
+    entry_point="write_histogram",
+    report_time=True,
+)
+hist_write_shader.set_resource(2, histogram_bins_buffer)
+hist_write_shader.set_resource(3, histogram_line_buffer)
+
+current_image_index = 0
+
+
+def draw_imgui():
+    global current_image_index
+
+    imgui.new_frame()
+
+    imgui.set_next_window_size((300, 0), imgui.Cond_.always)
+    imgui.set_next_window_pos((0, 0), imgui.Cond_.always)
+
+    is_expand, _ = imgui.begin("Controls", None)
+
+    if is_expand:
+        # Image selection dropdown
+        changed, current_image_index = imgui.combo(
+            "Image", current_image_index, standard_images, len(standard_images)
+        )
+        if changed:
+            # Create new texture object
+            img = load_image(standard_images[current_image_index])
+            image_texture = gfx.Texture(
+                img,
+                dim=2,
+                usage=wgpu.TextureUsage.STORAGE_BINDING
+                | wgpu.TextureUsage.TEXTURE_BINDING,
+            )
+            # Update the image
+            image_object.geometry.grid = image_texture
+            # Update histogram
+            hist_calc_shader.set_resource(0, image_object.geometry.grid)
+            histogram_object.local.scale_x = image_texture.size[0] / (nbins - 1)
+
+        # imgui.text(f"Histogram computation time: {computation_time * 1000:.1f} ms")
+
+    imgui.end()
+    imgui.end_frame()
+    imgui.render()
+    return imgui.get_draw_data()
+
+
+# Create GUI renderer
+gui_renderer = ImguiRenderer(renderer.device, canvas)
+gui_renderer.set_gui(draw_imgui)
+
+
+def animate():
+    if hist_calc_shader.changed:
+        size = image_object.geometry.grid.size
+        hist_calc_shader.dispatch(
+            int(size[0] / 16 + 0.499999), int(size[1] / 16 + 0.499999)
+        )
+        hist_write_shader.dispatch(1)
```

Comment:
How guaranteed is it that this will complete before the next scene draw?

Reply:
I'm pretty sure that Wgpu manages this.

Reply:
Also note the `device._poll_wait()`. That will actually halt the CPU until the GPU finished its work. We only do that when we want to measure the GPU's spent time.